lib/dns-manager.js

```diff
@@ -0,0 +1,318 @@
+                        }
+                    })
+                } else {
+                    dns.save(err => {
+                        if (err)
+                            debug('[%s] Error updating dns server configuration: %O', dns.id, err);
+                        else {
+                            debug('[%s] Deleted owner %s from dns server', dns.id, entityId);
+                        }
+                    });
+                }
+            }
+        });
+    });
+}
+
+createDnsServer = function (entityId, entityDomains, priority) {
+    let newId = randomstring.generate({
+        length: config.id_length,
+        capitalization: 'lowercase'
+    });
+    let dns = new Dns({
+        id: newId,
+        owner: [entityId],
+        domains: entityDomains,
+        priority: priority
+    });
+
+    debug('[%s] Saving new DNS Server configuration to database: %O', newId, dns);
+
+    dns.save((err, dns) => {
+        if (err) {
+            debug('[%s] Could not save new DNS Server configuration to database: %O', newId, err);
+        } else {
+            debug('[%s] Saved new DNS Server configuration to database', newId);
+            buildNewDnsServer(dns.id)
+                .then(() => {
+                    startNewDnsServer(dns.id);
+                });
+        }
+    })
+}
+
+buildNewDnsServer = function (dnsId) {
+    return new Promise((fulfill, reject) => {
+        let dnsConfigPath = path.join(config.fs.dns, dnsId);
+        Dns.findOne({id: dnsId}, (err, dns) => {
+            if (err) {
+                debug('[%s] Error searching in database: %O', dnsId, err);
+                reject(err);
+            } else if (!dns) {
+                debug('[%s] No DNS Server configuration found!', dnsId);
+                reject("No DNS Server configuration found!")
+            } else {
+                fse.mkdirsSync(dnsConfigPath);
+                fse.writeFileSync(dnsConfigPath + '/Dockerfile', config.dns.dockerfile);
+                fse.writeFileSync(dnsConfigPath + '/dnsmasq.conf', config.dns.dnsmasq.default);
+                addDomainsToDns(dnsConfigPath + '/dnsmasq.conf', dns.domains)
+                    .then(() => {
+                        let tarballFileName = path.join(config.payload.tarball.tmpdir, dnsId + '.tar');
+                        let tarballFile = fse.createWriteStream(tarballFileName);
+
+                        let archive = archiver('tar', {
+                            gzip: config.payload.tarball.gzip,
+                            gzipOptions: config.payload.tarball.gzipOptions,
+                            statConcurrency: config.payload.tarball.statConcurrency
+                        });
+
+                        archive.on('end', function () {
+                            debug('[%s] Packing payload to file %s completed (%s total bytes)', dnsId, tarballFileName, archive.pointer());
+
+                            docker.buildImage(tarballFileName, {t: dns.id}, (error, output) => {
```

Comment:
How quick is this usually? (I imaging it to be fast enough that you can just wait for it as a user)